### PR TITLE
Don't call local_storage_bytes in None case #320

### LINKED ISSUE #320: Client CLI Crashes When Printing Keys after Remote Gen Operation
The client CLI crashes with the following set of commands. Assume you are starting from a new server. Client CLI session:
```shell
| reg hackerman 1234
Logged in to hackerman
> ls
User not found
> g    
Stored: 1
> ls
name: 1
key_id: 90dee2b5b8acf5c617b7960b25c04f37cf02ef1037bafc3cb328f5b28499e5b4
data: Arbitrary Key - be93c1c792433778006c10e3359b99c47db11a0b299d8c6a163618c4ab9b5c1e


> remote-generate
Stored: 2
> ls
name: 2
Error converting LocalStorage to bytes
```

Followed by this backtrace:
```
   0:     0x557262671f2d - std::backtrace_rs::backtrace::libunwind::trace::h9135f25bc195152c
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x557262671f2d - std::backtrace_rs::backtrace::trace_unsynchronized::h015ee85be510df51
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x557262671f2d - std::sys_common::backtrace::_print_fmt::h5fad03caa9652a2c
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x557262671f2d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2b42ca28d244e5c7
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x55726269485c - core::fmt::write::h401e827d053130ed
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/fmt/mod.rs:1198:17
   5:     0x55726266cc71 - std::io::Write::write_fmt::hffec93268f5cde32
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/io/mod.rs:1672:15
   6:     0x5572626736f5 - std::sys_common::backtrace::_print::h180c4c706ee1d3fb
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x5572626736f5 - std::sys_common::backtrace::print::hd0c35d18765761c9
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x5572626736f5 - std::panicking::default_hook::{{closure}}::h1f023310983bc730
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:295:22
   9:     0x557262673411 - std::panicking::default_hook::h188fec3334afd5be
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:314:9
  10:     0x557262673c86 - std::panicking::rust_panic_with_hook::hf26e9d4f97b40096
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:698:17
  11:     0x557262673b77 - std::panicking::begin_panic_handler::{{closure}}::hfab912107608087a
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:588:13
  12:     0x557262672424 - std::sys_common::backtrace::__rust_end_short_backtrace::h434b685ce8d9965b
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x5572626738a9 - rust_begin_unwind
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
  14:     0x557261c07fc3 - core::panicking::panic_fmt::ha6dc7f2ab2479463
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
  15:     0x55726266be21 - std::io::stdio::print_to::h25dd002df6240e93
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/io/stdio.rs:1015:9
  16:     0x55726266be21 - std::io::stdio::_print::h235f89a7e7e6b2bb
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/io/stdio.rs:1027:5
  17:     0x557261c70b47 - lock_keeper_client_cli::storage::UserStore::list::hee7d61acc7abeaa4
                               at /home/omar/Programming/key-mgmt/lock-keeper-client-cli/src/storage.rs:192:13
  18:     0x557261d81eef - lock_keeper_client_cli::storage::Storage::list::h8f73a252ee7f23ff
                               at /home/omar/Programming/key-mgmt/lock-keeper-client-cli/src/storage.rs:94:9
  19:     0x557261c5ab53 - lock_keeper_client_cli::command::Command::execute::{{closure}}::h49e251ab34d6a046
                               at /home/omar/Programming/key-mgmt/lock-keeper-client-cli/src/command.rs:212:17
  20:     0x557261db0729 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h390548121572b262
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
  21:     0x557261d801dc - lock_keeper_client_cli::app::run::{{closure}}::h34566f5daa02ae5f
                               at /home/omar/Programming/key-mgmt/lock-keeper-client-cli/src/app.rs:23:52
  22:     0x557261dafc69 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1a1faa37774674ba
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
  23:     0x557261d86535 - lock_keeper_client_cli::main::{{closure}}::hd49627a30ba91aca
                               at /home/omar/Programming/key-mgmt/lock-keeper-client-cli/src/main.rs:15:39
  24:     0x557261db1b5b - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h76800b1a25d39226
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/future/mod.rs:91:19
  25:     0x557261ce37e0 - tokio::park::thread::CachedParkThread::block_on::{{closure}}::h03d7d203e1d7bfbd
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/park/thread.rs:267:54
  26:     0x557261d703a0 - tokio::coop::with_budget::{{closure}}::ha24267122963054d
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/coop.rs:102:9
  27:     0x557261d08bf1 - std::thread::local::LocalKey<T>::try_with::h49994f2063d91480
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:445:16
  28:     0x557261d0894d - std::thread::local::LocalKey<T>::with::h2dd7734adb3f1512
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:421:9
  29:     0x557261ce3676 - tokio::coop::with_budget::h83d753ceec89d37f
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/coop.rs:95:5
  30:     0x557261ce3676 - tokio::coop::budget::ha30c942d91b04357
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/coop.rs:72:5
  31:     0x557261ce3676 - tokio::park::thread::CachedParkThread::block_on::h5e86dd4318397580
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/park/thread.rs:267:31
  32:     0x557261d9c3a9 - tokio::runtime::enter::Enter::block_on::hd935eb72dc56e39c
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/runtime/enter.rs:152:13
  33:     0x557261cfdf3c - tokio::runtime::scheduler::multi_thread::MultiThread::block_on::hfd7839d0504389cd
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/runtime/scheduler/multi_thread/mod.rs:79:9
  34:     0x557261d9f670 - tokio::runtime::Runtime::block_on::hdcc015d2784796a2
                               at /home/omar/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.21.2/src/runtime/mod.rs:492:44
  35:     0x557261d44f55 - lock_keeper_client_cli::main::h6f0fcbe7f5d8d03b
                               at /home/omar/Programming/key-mgmt/lock-keeper-client-cli/src/main.rs:15:5
  36:     0x557261c442eb - core::ops::function::FnOnce::call_once::h01df62ed2c522c38
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ops/function.rs:248:5
  37:     0x557261c7df7e - std::sys_common::backtrace::__rust_begin_short_backtrace::ha1e492abda9a3eb9
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/sys_common/backtrace.rs:122:18
  38:     0x557261c96631 - std::rt::lang_start::{{closure}}::h57899d5dbf430576
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/rt.rs:145:18
  39:     0x5572626671fe - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hcdfee62722e5e4b8
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ops/function.rs:280:13
  40:     0x5572626671fe - std::panicking::try::do_call::h84ca51609826746f
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:492:40
  41:     0x5572626671fe - std::panicking::try::hd58075e533b8e0cb
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:456:19
  42:     0x5572626671fe - std::panic::catch_unwind::h1ebac24d83cb6ce2
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panic.rs:137:14
  43:     0x5572626671fe - std::rt::lang_start_internal::{{closure}}::h0145388a1edd1640
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/rt.rs:128:48
  44:     0x5572626671fe - std::panicking::try::do_call::h7630182e534a0a32
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:492:40
  45:     0x5572626671fe - std::panicking::try::h05b6544f0c6331dc
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:456:19
  46:     0x5572626671fe - std::panic::catch_unwind::h77b2ba8fd3309f34
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panic.rs:137:14
  47:     0x5572626671fe - std::rt::lang_start_internal::h6612c8a7a6861b8b
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/rt.rs:128:20
  48:     0x557261c96600 - std::rt::lang_start::hac3f9c2aa793335c
                               at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/rt.rs:144:17
  49:     0x557261d44ffc - main
  50:     0x7fb62532ed90 - __libc_start_call_main
                               at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  51:     0x7fb62532ee40 - __libc_start_main_impl
                               at ./csu/../csu/libc-start.c:392:3
  52:     0x557261c08315 - _start
  53:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by vidyaap**
Closes #320 

This was on me! I had the conversion to bytes outside of the match so it was trying to convert blank key material into bytes.

Also addresses part of #317 to move the audit event log check in the e2e tests to the end.

### CODE REVIEW

File: lock-keeper-tests/src/end_to_end.rs [hunk 1]
```diff
@@ -138,27 +138,28 @@ impl Test {
...
                             if expected_string != error_string {
                                 anyhow::bail!("Incorrect error. expected {expected_string}; got {error_string}")
                             }
                         }
-                        None => anyhow::bail!("Unexpected error"),
+                        None => anyhow::bail!("Unexpected error {error_string}"),
                     }
+                    self.check_audit_events(&self.config, EventStatus::Failed, &op)
+                        .await?;
                 }
```
Comment (gatoWololo) [inline lock-keeper-tests/src/end_to_end.rs:163]: I'm finding this logic here a bit hard to follow. I cleaned it up a bit by getting rid of the intermediate string conversions:
```rust
match outcome {
                Ok(()) => match &expected_outcome.expected_error {
                    Some(err) => {
                        anyhow::bail!("Unexpected success, expected: {err:?}")
                    }
                    None => {
                        self.check_audit_events(&self.config, EventStatus::Successful, &op)
                            .await
                    }
                }?,
                Err(e) => {
                    match &expected_outcome.expected_error {
                        Some(expected) => {
                            if expected.to_string() != e.to_string() {
                                anyhow::bail!("Incorrect error. expected {expected:?}; got {e:?}")
                            }
                        }
                        None => anyhow::bail!("Unexpected error {e}"),
                    }
                    self.check_audit_events(&self.config, EventStatus::Failed, &op)
                        .await?;
                }
            }
```

File: lock-keeper-tests/src/end_to_end.rs [hunk 2]
```diff
@@ -138,27 +138,28 @@ impl Test {
...
-                            let error_string = e.to_string();
                             if expected_string != error_string {
                                 anyhow::bail!("Incorrect error. expected {expected_string}; got {error_string}")
                             }
                         }
-                        None => anyhow::bail!("Unexpected error"),
+                        None => anyhow::bail!("Unexpected error {error_string}"),
                     }
+                    self.check_audit_events(&self.config, EventStatus::Failed, &op)
+                        .await?;
```
Comment (gatoWololo) [inline lock-keeper-tests/src/end_to_end.rs:162]: It looks like this line is only ever called in the Some arm of the match statement, when the `expected_string == error_string`. This isn't obvious from the logic. We should bring that into the match arm:
```rust
Err(e) => match &expected_outcome.expected_error {
                    Some(expected) => {
                        if expected.to_string() != e.to_string() {
                            anyhow::bail!("Incorrect error. expected {expected:?}; got {e:?}")
                        }
                        self.check_audit_events(&self.config, EventStatus::Failed, &op)
                            .await?;
                    }
                    None => anyhow::bail!("Unexpected error {e}"),
                },
```

Comment (gatoWololo) [reply lock-keeper-tests/src/end_to_end.rs:162]: Then I collapsed the match statements to try making it more clear what is going on:
```rust
match (outcome, expected_outcome.expected_error) {
                (Ok(()), Some(unexpected_err)) => {
                    anyhow::bail!("Unexpected success, expected: {:?}", unexpected_err);
                }
                (Ok(()), None) => {
                    self.check_audit_events(&self.config, EventStatus::Successful, &op)
                        .await?;
                }
                (Err(e), Some(expected_err)) => {
                    if expected_err.to_string() != e.to_string() {
                        anyhow::bail!("Incorrect error. expected {expected_err:?}; got {e:?}");
                    }
                    self.check_audit_events(&self.config, EventStatus::Failed, &op)
                        .await?;
                }
                (Err(e), None) => anyhow::bail!("Unexpected error {e}"),
            }
```

(Final version)

**Review by gatoWololo — CHANGES_REQUESTED** [review]
The PR looks good. I left some comments for refactoring the nested matching, which I think makes it more obvious what is going on. I tried to post my refactoring steps, so hopefully it is clear how I arrived at the final version. Instead of just posting the final version. Lmk if you have questions!

File: lock-keeper-tests/src/end_to_end.rs (hunk 2 above)
Comment (picklenerd) [reply lock-keeper-tests/src/end_to_end.rs:162]: Oh yeah I see it now, since the other branch returns an error.

Comment (vidyaap) [reply lock-keeper-tests/src/end_to_end.rs:162]: Thanks for this breakdown!! I hate to say it though, but the e2e test refactor PR (#325) gets rid of the match anyway 😅

Comment (gatoWololo) [reply lock-keeper-tests/src/end_to_end.rs:162]: After talking to Jordan, I am rewriting the Display function as `local_storage_bytes` is broken. So I actually don't think we need anything from this PR? :grimacing:

Comment (vidyaap) [reply lock-keeper-tests/src/end_to_end.rs:162]: lol no worries, if that's definite, I can just close this PR.

Comment (gatoWololo) [reply lock-keeper-tests/src/end_to_end.rs:162]: Yeah, that sounds good! Thanks.

**Comment by vidyaap** [comment]
Closing this because the CLI issue and the E2E issue are being fixed in #300 and #313 respectively.
